vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs

```diff
@@ -0,0 +1,223 @@
+
+    member this.BackgroundRequests = bgRequests
+    
+    /// Unittestable complement to LanguageServce.CreateSource
+    member this.CreateSource(buffer:IVsTextLines) : IFSharpSource =
+    
+        // Each time a source is created, also verify that the IProjectSite has been initialized to listen to changes to the project.
+        // We can't listen to OnProjectLoaded because the language service is not guaranteed to be loaded when this is called.
+        let filename = VsTextLines.GetFilename buffer
+        let rdt = this.ServiceProvider.RunningDocumentTable
+        let result = VsRunningDocumentTable.FindDocumentWithoutLocking(rdt,filename)
+        match result with 
+        | Some(hier,_) -> 
+            match hier with 
+            | :? IProvideProjectSite as siteProvider ->
+                let site = siteProvider.GetProjectSite()
+                site.AdviseProjectSiteChanges(FSharpCommonConstants.FSharpLanguageServiceCallbackName, 
+                                              new AdviseProjectSiteChanges(fun () -> this.OnProjectSettingsChanged(site))) 
+                site.AdviseProjectSiteCleaned(FSharpCommonConstants.FSharpLanguageServiceCallbackName, 
+                                              new AdviseProjectSiteChanges(fun () -> this.OnProjectCleaned(site))) 
+            | _ -> 
+                // This can happen when the file is in a solution folder or in, say, a C# project.
+                ()
+        | _ ->
+            // This can happen when renaming a file from a different language service into .fs or fsx.
+            // This naturally won't have an associated project.
+            ()
+        
+        // Create the source and register file change callbacks there.       
+        let source = this.SourceFactory(buffer)
+        this.ProjectSitesAndFiles.SetSource(buffer, source)
+        source
+    
+    // For each change in dependency files, notify the language service of the change and propagate the update
+    interface IDependencyFileChangeNotify with
+        member this.DependencyFileCreated projectSite = 
+            // Invalidate the configuration if we notice any add for any DependencyFiles 
+            let checkOptions = ProjectSitesAndFiles.GetProjectOptionsForProjectSite(projectSite, "")
+            this.FSharpChecker.InvalidateConfiguration(checkOptions)
+
+        member this.DependencyFileChanged (filename) = 
+            this.NotifyFileTypeCheckStateIsDirty filename
+
+
+    /// Do OnIdle processing for the whole language service. dirtyForTypeCheckFiles can be set by events 
+    /// raised on the background compilation thread.  
+    member this.OnIdle() = 
+        for file in dirtyForTypeCheckFiles do
+            let rdt = this.ServiceProvider.RunningDocumentTable
+            match this.ProjectSitesAndFiles.TryGetSourceOfFile(rdt, file) with
+            | Some source -> source.RecordChangeToView()
+            | None ->  ()
+        dirtyForTypeCheckFiles <- Set.empty
```

Comment:
If `dirtyForTypeCheckFiles` can be set on the background thread, is there a race condition here?